keps/sig-cluster-lifecycle/draft-20180130-kubeadm-join-master.md

```diff
@@ -0,0 +1,448 @@
+flow .
+
+## Graduation Criteria
+
+* To create a periodic E2E test that bootstraps an HA cluster with kubeadm
+  and exercise the dynamic bootstrap workflow
+* To ensure upgradability of HA clusters (possibly with another E2E test)
+* To document the kubeadm support for HA in kubernetes.io
+
+## Implementation History
+
+* original HA proposals [#1](https://goo.gl/QNtj5T) and [#2](https://goo.gl/C8V8PV)
+* merged [Kubeadm HA design doc](https://goo.gl/QpD5h8)
+* HA prototype [demo](https://goo.gl/2WLUUc) and [notes](https://goo.gl/NmTahy)
+* [PR #58261](https://github.com/kubernetes/kubernetes/pull/58261)
+
+## Drawbacks
+
+This proposal provide support for a single, well defined HA scenario.
+While this is considered a sustainable approach to the complexity of HA in Kubernetes,
+the limited scope of this proposal could be negatively perceived by final users.
+
+## Alternatives
+
+1)  Execute `kubeadm init` on many nodes
+
+The approach based on execution of `kubeadm init` on each master was considered as well,
+but not considered because this approach seems to have several draw backs:
+
+*  There is no real control on parameters passed to `kubeadm init` executed on secondary masters,
+    and this can lead to unpredictable inconsistent configurations.
+*  The init sequence for secondary master won't go through the TLS boostrap process,
+    and this can be perceived security concern.
+*  The init sequence executes a lot of steps with are un-necessary on a secondary master;
+    now those steps are mostly indepondent, so basically now no warm is done by executing
+    them two or three times. Nevertheless to mantain this contract in future could be complex.
+
+2) Allow HA configurations with `--advertise-address` equal to the master ip address
+(and adding the IP/DNS of the external load balancer as an additional apiServerCertSANs).
+
+After some testing, this option was considered too complex/not
+adequate for the initial set up of the desired `kubeadm join --master` dynamic workflow;
+this can be better explained by looking at two implementation based on this option:
+
+* [kubernetes the hard way](https://github.com/kelseyhightower/kubernetes-the-hard-way)
+   uses the IP address of all master nodes for creating a new the API server
+   serving certificate before bootstrapping the cluster, but this approach
+   can't be used if considering the desired dynamic workflow.
+
+* [Creating HA cluster with kubeadm](https://kubernetes.io/docs/setup/independent/high-availability/)
+  uses a different API server serving certificate for each master, and this
+  could increases the complexity of the first implementation because:
+  * the `kubeadm join --master` flow have to generate different certificates for
+    each master node.
+  * self-hosting control plane, should be adapted to mount different certificates
+    for each master.
+  * bootstrap check pointing should checkpoint be designed to checkpoin a different
```

Comment:
s/bootstrap check pointing should checkpoint be designed to checkpoin/bootstrap check pointing should be designed to checkpoint/